lib/puppet/provider/package/gem.rb

```diff
@@ -15,7 +15,17 @@
 
   has_feature :versionable, :install_options, :uninstall_options
 
-  commands :gemcmd => "gem"
+  if Puppet.features.microsoft_windows?
+    # Puppet on Windows prepends its paths to PATH. One of those paths includes Puppet's vendored ruby.
+    # That results in the which() method in self.command(name) in lib/puppet/provider.rb always returning Puppet's gem command.
+    # To prevent this, call which("gem") with a PATH without Puppet's paths, and set gemcmd to the full-path result. 
+    path_array = Puppet::Util.get_env("PATH").split(File::PATH_SEPARATOR)
+    path_array.reject! {|p| p =~ /puppet/i}
```

Comment:
Thanks, I made an assumption that users might change `INSTALLDIR` from the default of `C:\Program Files\Puppet Labs\Puppet` but not change that last `Puppet` directory.

Since we already find and set RUBY_DIR, could we reuse it as follows?

```
if Puppet.features.microsoft_windows?
  ruby_bin_dir = Puppet::Util.get_env('RUBY_DIR') + '\bin'
  path_array = Puppet::Util.get_env('PATH').split(File::PATH_SEPARATOR)
  path_array.reject! {|p| p == ruby_bin_dir}
  path_without_puppet_ruby_bin_dir = path_array.join(File::PATH_SEPARATOR)
  commands :gemcmd => Puppet::Util.withenv({:PATH => path_without_puppet_ruby_bin_dir}, :windows) { which('gem') }
else
  commands :gemcmd => 'gem'
end
```